cpp/dolfinx/fem/assembler.cpp

```diff
@@ -23,6 +23,57 @@
 using namespace dolfinx;
 using namespace dolfinx::fem;
 
+namespace
+{
+
+#ifdef PETSC_USE_64BIT_INDICES
+const std::function<int(std::int32_t, const std::int32_t*, std::int32_t,
+                        const std::int32_t*, const PetscScalar*)>
+make_petsc_lambda(Mat& A, std::vector<PetscInt>& storage_p)
+{
+
+  const std::function<int(std::int32_t, const std::int32_t*, std::int32_t,
+                          const std::int32_t*, const PetscScalar*)>
+      f = [&A, &storage_p](std::int32_t nrow, const std::int32_t* rows,
+                           std::int32_t ncol, const std::int32_t* cols,
+                           const PetscScalar* y) {
+        storage_p.resize(nrow + ncol);
+        std::copy(rows, rows + nrow, storage_p.begin());
+        std::copy(cols, cols + ncol, storage_p.begin() + nrow);
+        const PetscInt *rows1 = storage_p.data(), *cols1 = rows1 + nrow;
+        PetscErrorCode ierr
+            = MatSetValuesLocal(A, nrow, rows1, ncol, cols1, y, ADD_VALUES);
+#ifdef DEBUG
+        if (ierr != 0)
+          la::petsc_error(ierr, __FILE__, "MatSetValuesLocal");
+#endif
+        return 0;
+      };
+  return f;
+}
+#else
+const std::function<int(std::int32_t, const std::int32_t*, std::int32_t,
+                        const std::int32_t*, const PetscScalar*)>
+make_petsc_lambda(Mat& A, std::vector<PetscInt>&)
```

Comment:
See above comments on `A` vs `&A`.

Reply:
I am a bit torn about `A` vs `&A`. Practically it makes no difference - a reference to a pointer, versus copying the pointer, right? But if `Mat` was an object, then this would show that it is not `const` - we are changing it in the lambda. I understand your point, but I think the reference shows the intent...

Reply:
But we're not changing `A` - we're changing what `A` points to. We overcomplicated interfacing to PETSc in the past because we treated `Mat` as an object rather than as a pointer.